titus-server-master/src/main/java/com/netflix/titus/master/supervisor/service/DefaultSupervisorOperations.java

```diff
@@ -0,0 +1,145 @@
+import javax.inject.Singleton;
+
+import com.netflix.governator.LifecycleManager;
+import com.netflix.titus.common.runtime.SystemLogEvent;
+import com.netflix.titus.common.runtime.TitusRuntime;
+import com.netflix.titus.common.util.CollectionsExt;
+import com.netflix.titus.common.util.rx.ObservableExt;
+import com.netflix.titus.common.util.tuple.Pair;
+import com.netflix.titus.master.supervisor.model.MasterInstance;
+import com.netflix.titus.master.supervisor.model.MasterInstanceFunctions;
+import com.netflix.titus.master.supervisor.model.MasterState;
+import com.netflix.titus.master.supervisor.model.event.MasterInstanceRemovedEvent;
+import com.netflix.titus.master.supervisor.model.event.MasterInstanceUpdateEvent;
+import com.netflix.titus.master.supervisor.model.event.SupervisorEvent;
+import com.netflix.titus.runtime.endpoint.metadata.CallMetadata;
+import com.netflix.titus.runtime.endpoint.metadata.CallMetadataUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import rx.Observable;
+
+@Singleton
+public class DefaultSupervisorOperations implements SupervisorOperations {
+
+    private static final Logger logger = LoggerFactory.getLogger(DefaultSupervisorOperations.class);
+
+    private static final long GRACEFUL_SHUTDOWN_TIMEOUT_MS = 10_000;
+
+    private final LifecycleManager lifecycleManager;
+    private final MasterMonitor masterMonitor;
+    private final TitusRuntime titusRuntime;
+
+    @Inject
+    public DefaultSupervisorOperations(LifecycleManager lifecycleManager,
+                                       MasterMonitor masterMonitor,
+                                       TitusRuntime titusRuntime) {
+        this.lifecycleManager = lifecycleManager;
+        this.masterMonitor = masterMonitor;
+        this.titusRuntime = titusRuntime;
+    }
+
+    @Override
+    public List<MasterInstance> getMasterInstances() {
+        return masterMonitor.observeMasters().take(1).toBlocking().first();
+    }
+
+    @Override
+    public Optional<MasterInstance> findMasterInstance(String instanceId) {
+        return getMasterInstances().stream()
+                .filter(i -> i.getInstanceId().equals(instanceId))
+                .findFirst();
+    }
+
+    @Override
+    public MasterInstance getMasterInstance(String instanceId) {
+        return findMasterInstance(instanceId).orElseThrow(() -> SupervisorServiceException.masterInstanceNotFound(instanceId));
+    }
+
+    @Override
+    public Optional<MasterInstance> findLeader() {
+        return getMasterInstances().stream().filter(i -> MasterState.isLeader(i.getStatus().getState())).findFirst();
+    }
+
+    @Override
+    public Observable<SupervisorEvent> events() {
+        return masterMonitor.observeMasters()
+                .compose(ObservableExt.mapWithState(Collections.emptyMap(), this::buildEventList))
+                .flatMap(Observable::from);
+    }
+
+    @Override
+    public void restartMasterInstance(CallMetadata callMetadata) {
```

Comment:
it looks like this still needs to be renamed based on the discussion on Netflix/titus-api-definitions#55

Reply:
Yes, I have renamed this to `stopBeingLeader`.